### ISSUE TITLE
Can't resolve 'module' error when importing npm module for the brower + wasm compile error

### ISSUE THREAD

**Issue by frading** [elalish/manifold#365]
Hello,

As I mentioned in the [Users of Manifold discussion](https://github.com/elalish/manifold/discussions/340), I'm starting to integrate Manifold in [Polygonjs](https://github.com/polygonjs/polygonjs). But I'm quickly facing issues, as importing the library generates the following error:

```
ERROR in ./node_modules/manifold-3d/manifold.js 8:8370-8386
Module not found: Error: Can't resolve 'module' in '.../polygonjs/node_modules/manifold-3d'
resolve 'module' in '.../polygonjs/node_modules/manifold-3d'
Parsed request is a module
using description file: .../polygonjs/node_modules/manifold-3d/package.json (relative path: .)
Field 'browser' doesn't contain a valid alias configuration
```

It seems to come from `await import('module')` inside `node_modules/manifold-3d/manifold.js`. So I can temporarily work around it by removing the whole block inside `if (ENVIRONMENT_IS_NODE) {...}`. (as I'm building for the browser, not for node).

Then the app builds, but loading the wasm file generates a second error:

```
Aborted(CompileError: WebAssembly.instantiate(): expected magic word 00 61 73 6d, found 3c 21 44 4f @+0). Build with -sASSERTIONS for more info.
```

And this is unfortunately not an error I'm able to debug.

So this brings me 2 questions:

1 - Would it be possible to have a build in the npm module that is only for the browser and does not include nodejs specific instructions? Alternatively, could you suggest modifications to my webpack config that could solve this?

2 - Is there anything I should change to load the .wasm without the error above?

This issue may need to be split into 2 for clarity, feel free to let me know if so, I'll happily do it.

You can see my attempt to integrate manifold in the [integration-manifold branch of polygonjs](https://github.com/polygonjs/polygonjs/tree/integration-manifold), and specifically [this file](https://github.com/polygonjs/polygonjs/blob/integration-manifold/src/engine/nodes/sop/SDFBox.ts). You can also start it with `yarn && yarn start` and view any errors in the dev console when opening it on `http://localhost:8080/example.html?id=nodes/sop/SDFBox`.

Don't hesitate to let me know if you need more info from me

**Comment by pca006132** [comment]
This is weird, we do have tests running with nodejs in our CI:

https://github.com/elalish/manifold/blob/dca9cce5299f8829da79fada0a345beb615ad82c/.github/workflows/manifold.yml#L132

What is the version for your nodejs?

**Comment by frading** [comment]
Thank you for your quick reply.

I'm on `v16.15.1`, but to confirm, I'm trying to build for the browser, not for node.

And to help debug this, I've created a test repo https://github.com/frading/manifold-3d-test, where all I've done is:

- scaffold the project using vite with `yarn create vite`
- add a config that uses vite wasm plugin
- try to import manifold in https://github.com/frading/manifold-3d-test/blob/main/src/main.ts with:

```
import Module from 'manifold-3d'

async function loadManifold(){
  console.log(Module);
  const wasm = await Module()
  console.log(wasm);
  wasm.setup();
}
loadManifold()
```

And in this test, the error I get is different: `GET http://localhost:5173/node_modules/.vite/deps/manifold.wasm net::ERR_ABORTED 404 (Not Found)`

So webpack and vite both try and import it differently. I'll keep investigate to try and see how I can tweak those configs, but curious to hear what you think in the mean time.

Another suggestion that comes to mind, is that it would help import it if it was possible to give the url where to load the .wasm from to the `Module()` call. So it would be something like `Module({wasmUrl: '/my-custom-folder/manifold.wasm'})`. That would avoid having to configure the bundler to handle .wasm, which is always tricky.

**Comment by pca006132** [comment]
OK I think I misunderstood the problem. From emscripten documentation, it seems that you can modify the compiler args to build it for the web (https://emscripten.org/docs/getting_started/FAQ.html#can-i-build-javascript-that-only-runs-on-the-web) and bundle it as a single file which hopefully makes it easier for webpack/vite to work with (https://emscripten.org/docs/getting_started/FAQ.html#how-do-i-run-a-local-webserver-for-testing-why-does-my-program-stall-in-downloading-or-preparing)

**Comment by frading** [comment]
ah, I did not realise the current .wasm was only for node. Ok, thanks a lot for those links, I'll give that a go and report back.

**Comment by pca006132** [comment]
It should not be just for node, but I have not used frontend tools for a while and have no idea why it does not work with webpack. Please let us know your results, and if this works we can add it to the README.

**Comment by elalish** [comment]
Our npm library definitely works in the browser, simply look at manifoldcad.org, which uses it this way. From the file you link, the strangest part to me is that you await the module import and setup the WASM inside the `cook` function that does the work. Certainly you wouldn't want to do this if you call it multiple times - you should set up the WASM once and then use it from then on. See for instance how we do it here: https://github.com/elalish/manifold/blob/master/bindings/wasm/test/test.js

**Comment by frading** [comment]
> you should set up the WASM once and then use it from then on

Yes, you're totally right, what I linked to is just a preliminary test to try and load your library. Once I got that working, it'll be cached.

But for now, my attempts to load it, in both repos I've linked, using webpack or vite, have failed. And unless I'm missing something obvious, I'm doing it the same way as the file you just pointed to.

**Comment by jirihon** [comment]
Hi @frading, I'm successfully using Manifold with Vite in a SvelteKit project. You just need to point the `Module` function to the location of the manifold.wasm. See:

```js
import Module from 'manifold-3d';
import manifold_wasm from 'manifold-3d/manifold.wasm?url';
const manifold_module = await Module({
    locateFile: () => manifold_wasm,
});
manifold_module.setup();
```

The `?url` syntax in the import is [explicit URL import](https://vitejs.dev/guide/assets.html#explicit-url-imports). Not sure if Webpack has something similar.

See Manifold running in an example Svelte project using Vite at https://stackblitz.com/edit/manifold-vite?file=src%2FApp.svelte&terminal=dev

If you use TypeScript, it will complain that the `Module` function does not have any argument. This is an inaccuracy in the TypeScript definitions of the Manifold's JS bindings.

So, currently, Manifold npm package is not usable without this extra knowledge with Vite (and Webpack). I don't know yet how this should be done properly. I'll try to find some other package that is also using wasm and is working out-of-the-box with Vite, for inspiration. But definitely, a least the TypeScript definitions could be fixed easily, so it's clear for the TypeScript users there is the `locateFile` parameter that can be set (and should be set when using bundlers).

**Comment by jirihon** [comment]
It seems that npm packages using wasm modules encouters the exact same problem. And providing the URL to the wasm file to the module load function is the best option. For example, here is a relevant documentation for OpenCascade.js: [Configuring Your Bundler](https://ocjs.org/docs/getting-started/configure-bundler).
